plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevExceptions.kt

```diff
@@ -6,33 +6,63 @@ package software.aws.toolkits.jetbrains.services.amazonqFeatureDev
 import software.aws.toolkits.jetbrains.services.amazonq.RepoSizeError
 import software.aws.toolkits.resources.message
 
-open class FeatureDevException(override val message: String?, override val cause: Throwable? = null) : RuntimeException()
+/**
+ * FeatureDevException models failures from feature dev operations.
+ *
+ * - Each failure is annotated based on className, operation, and a short desc. Use the `reason()` and `reasonDesc()` members for instrumentation.
+ * - To throw an exception without modeling, throw FeatureDevException directly.
+ */
+open class FeatureDevException(override val message: String, open val operation: String, open val desc: String?, override val cause: Throwable? = null)
+    : RuntimeException() {
+    fun reason(): String {
```

Comment:
Not seeing this necessary. We use class name as reason. Why don't create explicit Error class?

```
class UploadCodeError(override val message: String? = message("amazonqFeatureDev.exception.upload_code"), override val cause: Throwable? = null) : FeatureDevException(message, cause)

```

Reply:
> Why don't create explicit Error class, and replace those unclear FeatureDevException?

There are several catch-alls currently throwing `FeatureDevException` using`apiError()`. In this changeset, we retain the existing behavior of these not being modeled more explicitly. In other words, 1) enumerated error cases should be as you describe with explicit subclasses, 2) non-enumerated error cases (i.e. catch-alls) throw `FeatureDevException`. I agree we can/should enumerate additional errors, but I think that's beyond the scope of this changeset.

Reply:
> Not seeing this necessary. We use class name as reason. 

I'm happy to remove if you prefer. I expanded on this in another comment, so I'll resolve this conversation. It's a tradeoff between explicit expectations for how to use the class (this is what I biased toward in this revision) vs reducing redundancy at the call sites, which call this directly in addition to the lookup for other classes.

What do you think?